File: routes/ocorrencia.py
"""Endpoint para registrar ocorrências enviadas pelo WhatsApp."""

from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, Field

from functions.save_ocorrencia import save_ocorrencia_texto
from functions.db_client import get_client_db

router = APIRouter()


class OcorrenciaRequest(BaseModel):
    """Modelo de entrada para ocorrência."""
    nomovtra: int = Field(..., description="Número do pedido/entrega (NOMOVTRA)")
    texto: str = Field(..., description="Descrição da ocorrência")
    usuario: str = Field(..., description="Identificação do usuário (CPF, CNPJ ou BOT)")


@router.post("/ocorrencia")
async def registrar_ocorrencia(
    req: OcorrenciaRequest,
    to_biz: str = Header(..., alias="x-whatsapp-number"),
) -> dict:
    """Grava a ocorrência na tabela TABMOVTRA_OCO do cliente."""
    try:
        cfg = get_client_db(to_biz)
        save_ocorrencia_texto(
            req.nomovtra, req.texto, req.usuario, cfg
        )
        return {"status": "ok"}
    except Exception as exc:  # pragma: no cover - falha inesperada
        raise HTTPException(status_code=500, detail=str(exc))

File: twilio/src/routes/ocorrencia.js
import express from 'express';
import logger from '../utils/logger.js';
import { getClientByNumber, getTenantPool } from '../services/db.js';
import { inserirOcorrencia } from '../services/ocoService.js';

const router = express.Router();

router.post('/ocorrencia', async (req, res) => {
  const toBiz = req.header('x-whatsapp-number') || req.body?.toBiz;
  const { nomovtra, texto, cpf } = req.body || {};

  if (!toBiz) {
    return res.status(400).json({ error: "x-whatsapp-number ausente" });
  }
  if (!nomovtra || !texto || !cpf) {
    return res
      .status(400)
      .json({ error: "Campos obrigatórios: nomovtra, texto, cpf" });
  }

  try {
    const client = await getClientByNumber(toBiz);
    if (!client) {
      return res.status(404).json({ error: "Cliente não encontrado" });
    }
    const pool = getTenantPool(client);
    const resultado = await inserirOcorrencia(nomovtra, texto, cpf, pool);
    return res.json({ status: "ok", ...resultado });
  } catch (err) {
    logger.error({ err: err?.message || err }, 'falha registrar ocorrência');
    return res.status(500).json({ error: 'Erro ao registrar ocorrência' });
  }
});

export default router;
